mqtt-level-store.js

```diff
@@ -19,38 +19,112 @@ function Store (options) {
   }
 }
 
+const zeroPadding = (function () {
+  // width and padding calculates only once
+  const width = parseInt(Number.MAX_SAFE_INTEGER).toString(10).length
+  const padding = new Array(width + 1).join('0')
+  return function (num) {
+    return (padding + num).slice(-width)
+  }
+})()
+
 Store.prototype.put = function (packet, cb) {
-  this._level.put('' + packet.messageId, packet, this._levelOpts, cb)
+  var date = zeroPadding(Date.now())
+  if (this._prevDate === date) {
+    ++this._sameDateCount
+  } else {
+    this._sameDateCount = 0
+    this._prevDate = date
+  }
+  date += zeroPadding(this._sameDateCount)
+
+  var that = this
+  this._level.get(
+    'packets~' + packet.messageId,
+    this._levelOpts,
+    function (err, _date) {
+      if (err) {
+        if (err.notFound) {
+          var cmd = [
+            {type: 'put', key: 'packets~' + packet.messageId, value: date},
+            {type: 'put', key: 'packet-by-date~' + date + '~' + packet.messageId, value: packet}
+          ]
+          that._level.batch(cmd, that._levelOpts, cb)
+        } else {
+          cb(err)
+        }
+      } else {
+        that._level.get(
+          'packet-by-date~' + _date + '~' + packet.messageId,
+          that._levelOpts,
+          function (err, _packet) {
+            if (err) return cb(err)
+            if (packet.cmd === _packet.cmd) {
+              cb()
+            } else {
+              that._level.put('packet-by-date~' + _date + '~' + packet.messageId, packet, that._levelOpts,
+                function () {
+                  cb()
+                }
+              )
+            }
+          })
+      }
+    })
   return this
 }
 
 Store.prototype.get = function (packet, cb) {
-  this._level.get('' + packet.messageId, this._levelOpts, cb)
+  var that = this
+  this._level.get(
+    'packets~' + packet.messageId,
+    this._levelOpts,
+    function (err, date) {
+      if (err) return cb(err)
+      that._level.get(
+        'packet-by-date~' + date + '~' + packet.messageId,
+        that._levelOpts,
+        function (err, packet) {
+          cb(err, packet)
+        })
```

Comment:
can you just pass `cb` to `_level.get()`?

Reply:
Yes, I will fix it.